source/brailleDisplayDrivers/freedomScientific.py

```diff
@@ -1,292 +1,466 @@
-	fbClose=getattr(fsbLib,'_fbClose@4')
-	fbConfigure=getattr(fsbLib, '_fbConfigure@8')
-	fbGetDisplayName=getattr(fsbLib, "_fbGetDisplayName@12")
-	fbGetFirmwareVersion=getattr(fsbLib,  "_fbGetFirmwareVersion@12")
-	fbBeep=getattr(fsbLib, "_fbBeep@4")
-
-FB_INPUT=1
-FB_DISCONNECT=2
-FB_EXT_KEY=3
-
-LRESULT=c_long
-HCURSOR=c_long
-
-appInstance=windll.kernel32.GetModuleHandleW(None)
-
-nvdaFsBrlWm=windll.user32.RegisterWindowMessageW(u"nvdaFsBrlWm")
-
-inputType_keys=3
-inputType_routing=4
-inputType_wizWheel=5
-
-# Names of freedom scientific bluetooth devices
-bluetoothNames = (
-	"F14", "Focus 14 BT",
-	"Focus 40 BT",
-	"Focus 80 BT",
-)
-
-keysPressed=0
-extendedKeysPressed=0
-@WNDPROC
-def nvdaFsBrlWndProc(hwnd,msg,wParam,lParam):
-	global keysPressed, extendedKeysPressed
-	keysDown=0
-	extendedKeysDown=0
-	if msg==nvdaFsBrlWm and wParam in (FB_INPUT, FB_EXT_KEY):
-		if wParam==FB_INPUT:
-			inputType=lParam&0xff
-			if inputType==inputType_keys:
-				keyBits=lParam>>8
-				keysDown=keyBits
-				keysPressed |= keyBits
-			elif inputType==inputType_routing:
-				routingIndex=(lParam>>8)&0xff
-				isRoutingPressed=bool((lParam>>16)&0xff)
-				isTopRoutingRow=bool((lParam>>24)&0xff)
-				if isRoutingPressed:
-					gesture=RoutingGesture(routingIndex,isTopRoutingRow)
-					try:
-						inputCore.manager.executeGesture(gesture)
-					except inputCore.NoInputGestureAction:
-						pass
-			elif inputType==inputType_wizWheel:
-				numUnits=(lParam>>8)&0x7
-				isRight=bool((lParam>>12)&1)
-				isDown=bool((lParam>>11)&1)
-				#Right's up and down are rversed, but NVDA does not want this
-				if isRight: isDown=not isDown
-				for unit in xrange(numUnits):
-					gesture=WizWheelGesture(isDown,isRight)
-					try:
-						inputCore.manager.executeGesture(gesture)
-					except inputCore.NoInputGestureAction:
-						pass
-		elif wParam==FB_EXT_KEY:
-			keyBits=lParam>>4
-			extendedKeysDown=keyBits
-			extendedKeysPressed|=keyBits
-		if keysDown==0 and extendedKeysDown==0 and (keysPressed!=0 or extendedKeysPressed!=0):
-			gesture=KeyGesture(keysPressed,extendedKeysPressed)
-			keysPressed=extendedKeysPressed=0
-			try:
-				inputCore.manager.executeGesture(gesture)
-			except inputCore.NoInputGestureAction:
-				pass
-		return 0
-	else:
-		return windll.user32.DefWindowProcW(hwnd,msg,wParam,lParam)
-
-nvdaFsBrlWndCls=WNDCLASSEXW()
-nvdaFsBrlWndCls.cbSize=sizeof(nvdaFsBrlWndCls)
-nvdaFsBrlWndCls.lpfnWndProc=nvdaFsBrlWndProc
-nvdaFsBrlWndCls.hInstance=appInstance
-nvdaFsBrlWndCls.lpszClassName=u"nvdaFsBrlWndCls"
-
-class BrailleDisplayDriver(braille.BrailleDisplayDriver,ScriptableObject):
-
-	name="freedomScientific"
+import hwIo
+import serial
+
+
+BAUD_RATE = 57600
+PARITY = serial.PARITY_NONE
+
+#: Model names and number of cells
+MODELS = {
+	"Focus 14": 14,
+	"Focus 40": 40,
+	"Focus 44": 44,
+	"Focus 70": 70,
+	"Focus 80": 80,
+	"Focus 84": 84,
+	"pm display 20": 20,
+	"pm display 40": 40,
+}
+
+# Packet types
+FS_PKT_QUERY = "\x00"

```

Comment:
These will become unicode strings when moving to python 3. Do you mind giving all these the bytes literal explicitly?

Reply:
Sure, however I think we need to go through the whole string handling path for braille drivers. Since it is low level communication, we don't want any unicode in there, just bytes and explicit conversion when it enters the application (e.g. when setting self._model in this case). I'd rather prefer to fix this now everywhere in this driver or not do it at all. Marking the packet types as bytestrings might imply this is all written with Py3 unicode/bytes in mind, which it is not at this stage.